source/common/http/http2/codec_impl.cc

```diff
@@ -138,20 +138,14 @@ void ConnectionImpl::StreamImpl::encodeTrailersBase(const HeaderMap& trailers) {
 
 void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadata_map_vector) {
   ASSERT(parent_.allow_metadata_);
-
-  getMetadataEncoder().createPayload(metadata_map_vector);
-
-  // Estimates the number of frames to generate, and breaks the while loop when the size is reached
-  // in case submitting succeeds and packing fails, and we don't get error from packing.
-  const size_t frame_count = metadata_encoder_->frameCountUpperBound();
-  size_t count = 0;
-  // Keep submitting extension frames if there is payload left in the encoder.
-  while (metadata_encoder_->hasNextFrame() && count++ <= frame_count) {
-    submitMetadata();
-    parent_.sendPendingFrames();
+  MetadataEncoder& metadata_encoder = getMetadataEncoder();
+  if (!metadata_encoder.createPayload(metadata_map_vector)) {
+    return;
   }
-
-  ASSERT(!metadata_encoder_->hasNextFrame());
+  for (uint8_t flags : metadata_encoder.payloadFrameFlags()) {
```

Comment:
naming nit: shouldn't this be 
for (flag : flags) rather than
for (flags : flags)?

Reply:
Each of these is a technically a byte of flags, even though only the one bit is ever set or read. I renamed payloadFrameFlags -> payloadFrameFlagBytes instead.